core/src/SYCL/Kokkos_SYCL.cpp

```diff
@@ -93,7 +93,7 @@ bool SYCL::impl_is_initialized() {
 
 void SYCL::impl_finalize() { Impl::SYCLInternal::singleton().finalize(); }
 
-void SYCL::fence() const { m_space_instance->m_queue->wait(); }
+void SYCL::fence() const { m_space_instance->m_queue->wait_and_throw(); }
```

Comment:
@dhollman says:
> Please elaborate on this change with a comment.

Reply:
`wait()` only reports synchronous errors through `SYCL` exceptions while `wait_and_throw` also passes asynchronous errors to the exception handler.

Reply:
To elaborate:  exceptions thrown in kernels will, on exit, call an async handler on the host associated with the queue.  The default one is supposed to report the exceptions and then `std::terminate` (I haven't tried this is a while, though).

Reply:
So we handle async exception by printing a message and aborting but synchronous error will yield an uncaught (by Kokkos) exception?

Reply:
Is that a conscious design decision?

Reply:
I am throwing now also for asynchronous errors.

Reply:
I still don't think this is consistent handling of synchronous vs asynchronous exceptions because you trade async SYCL exceptions for `std::runtime_error`

Reply:
`CUDA_SAFE_CALL` in the end also throws a `std::runtime_error`. So that behavior is at least consistent within `Kokkos`.
I am not quite sure what kind of alternative you are thinking about. There might be multiple asynchronous exceptions and I think we agreed on addressing all of them. If we just throw, we only have a chance to process the first one. Otherwise, as done here, we can collect/print all the errors and then throw in the end.

Reply:
I was thinking a try-catch block around `queue::wait_and_throw()`